[GeoMechanicsApplication] Re-enable settlement test using the C++ route

avdg81 opened the pull request:
**📝 Description**

Converted a _disabled_ C++ unit test to a Python test. It shouldn't be a C++ unit test, since it tests a complete workflow, not just a single component. The new Python test covers the so-called "C++ route" of the settlement workflow. In other words, it uses the exact same entry point into the code as D-GeoSettlement does.

Since the `KratosGeoSettlement` class owns a local `KratosGeoMechanicsApplication` instance, running this new test will deregister all GeoMechanicsApplication components (e.g. geo elements) when that object is destroyed. To avoid that any subsequent tests start failing, we need to reload the relevant Kratos applications at the end of the new test (in its `tearDown` method).

With these changes, there are now two Python tests that simulate the same problem (a settlement problem), but using different entry points in our code (i.e. the "C++ route" and the "Python route"). Both ways should yield the same results. That is asserted by these tests.

avdg81 committed:
Run `settlement_workflow.py` in subdirectory `python`

The test setup creates a clean test directory that contains copies of the Kratos input files.

avdg81 committed:
Renamed a Python test script and a test class

The new names now clearly indicate that the test covers the Python route.

avdg81 committed:
Added an incomplete Python test script for the C++ route

C++ classes `CustomWorkflowFactory` and `KratosGeoSettlement` are now accessible from Python code.

avdg81 committed:
Introduced an intermediate test fixture

This new class is intended to be used by the test case for the Python route as well as the one for the C++ route. It will keep common functionality.

avdg81 committed:
Added two new attributes to the base test fixture

One attribute keeps the number of stages to be analyzed, the other the names of the project parameters files.

avdg81 committed:
Renamed a Python test script

avdg81 committed:
Added a test fixture for the C++ route

- Make sure that the heap-allocated settlement API object is properly freed when the test has finished.
- Make sure that the GeoMechanicsApplication has been imported prior to running the Python route of the settlement analysis.
- The C++ route needs to be tested more thoroughly.

avdg81 committed:
Fixed an incorrect Python module name

avdg81 committed:
Also corrected an import statement

avdg81 committed:
Added more checks for the C++ route

avdg81 committed:
Add a lambda for the "should cancel" callback

avdg81 committed:
Attempt to fix test failures

avdg81 committed:
Use proper callback functions in the C++ route test

avdg81 committed:
Defined a map that contains the time steps to check

avdg81 committed:
Added time, output filename and expected displacements

These are currently only being used by the test case for the Python route.

avdg81 committed:
Extracted a method for checking the displacements

This method is now also used by the test case for the C++ route.

avdg81 committed:
Extracted a method for defining the expected displacements

avdg81 committed:
Defined methods for checking the nodal stresses

avdg81 committed:
Also the test case for the C++ route now checks the nodal stresses

avdg81 committed:
Updated the descriptions of the test fixtures

avdg81 committed:
Removed the C++ unit test for the settlement workflow

avdg81 committed:
Removed the now unused original output files

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/12865-re-enable-cpp-settlement-test

avdg81 committed:
Use `def_static` rather than `def` to bind static member functions

avdg81 committed:
Removed an unused `#include`

avdg81 committed:
Removed an unused test script

avdg81 committed:
Added a missing `tearDown` call of the base class

avdg81 committed:
Moved a function definition to another line in the file

avdg81 committed:
Don't use a loop counter when we don't need one

rfaasse commented on applications/GeoMechanicsApplication/custom_python/add_custom_utilities_to_python.h:
```diff
@@ -15,6 +15,7 @@
 #define KRATOS_ADD_UTILITIES_TO_PYTHON_H_INCLUDED
 
 // System includes
+#include <pybind11/functional.h>
```
For those interested, this include is needed such that python (lambda) functions can be converted to C++ ones (since we pass functions to the exported RunStage function)

rfaasse commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
 
-            actual_data = reader.read_output_from(result_file_name)
-            actual_nodal_values = reader.nodal_values_at_time("DISPLACEMENT", times_to_check[i], actual_data, node_ids)
+        super().tearDown()
```
To me it makes sense to first do the super().tearDown(), but I'm not too sure. One could also argue that the super should always be on the 'outside' (so before for the setup and after for the teardown). What are your thoughts?

avdg81 replied:
My thought process was the analogy with constructors and destructors. Anything that a constructor builds up is torn down in reverse order by the destructor. So my reasoning was that `setUp` first does what the base class prescribes (by calling `super().setUp()`) and then adds specific new things for that test fixture. In `tearDown` we move backwards: so first clean up anything specific and then the stuff maintained by the base class (by calling `super().tearDown()`).

I have checked the [documentation of Python `unittest`](https://docs.python.org/3/library/unittest.html), but I couldn't find any advice there. What are your thoughts about my approach?

rfaasse replied:
Agreed, I was indeed in two minds, but this seems more symmetrical (i.e. super 'sandwiches' the things the derived class does), so let's leave it as is

rfaasse commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
+        no_logging = lambda msg: None
+        no_progress_reporting = lambda fraction_done: None
+        no_progress_message = lambda msg: None
+        dont_cancel = lambda: False
```
Very nitpicky: 
```suggestion
        do_not_cancel = lambda: False
```

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
+    """
+    def tearDown(self):
+        # The `KratosGeoSettlement` instance used by this test removes all registered GeoMechanicsApplication
+        # components when it's destroyed. If no action is taken, any following tests will start to fail. It seems
```
Minor clarification as to why this fails the next tests
```suggestion
        # components when it's destroyed. If no action is taken, any following tests will start to fail due to components not being registered. It seems
```

avdg81 replied:
Thank you for your suggestion. I have accepted your clarification.

rfaasse commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
+                                 }]
+
+
+    def check_displacements(self):
```
Nice table-based implementation of what to check 👍

avdg81 replied:
Thanks.

rfaasse reviewed (dismissed):
Very nice we finally have the C++ test running in our pipelines, thanks for getting this done (even though there were quite specific issues on the way). I really like the table-based way of comparing the results! I don't have any blocking comments, just a few questions/minor suggestions

markelov208 commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
+        try:
+            shutil.rmtree(self.test_path)
+        except FileNotFoundError:
+            pass
```
Copilot suggests  
shutil.rmtree(self.test_path, ignore_errors=True)

avdg81 replied:
Thank you for your suggestion. It is clearer and more expressive. I have adopted your suggestion.

markelov208 commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -1,61 +1,195 @@
+
+        self.number_of_stages = 4
+        self.project_parameters_filenames = [f"ProjectParameters_stage{i+1}.json" for i in range(self.number_of_stages)]
+        input_filenames = self.project_parameters_filenames[:] + ["MaterialParameters.json", "test_model.mdpa"]
```
I think this is minor: Copilot suggests to remove [:]

avdg81 replied:
I'm wasn't sure here. The reason I've added `[:]` (i.e. slicing) is that it will make a deep copy. And then I add two more items to this copy. I just wanted to make sure that `self.project_parameters_filenames` and `input_filenames` don't point to the same list object.

But based on your suggestions, I have done a few tests. And I found that your suggestion also makes a deep copy of the first list. So I will adopt your suggestion. Thanks for pointing this out to me.

markelov208 replied:
Thank you for the study. Copilot wrote that its change will eliminate the copying but if it still copies. ;(

markelov208 reviewed (commented):
Hi Anne, thank you for the clear and well understandable code. I am still new, so I consulted Copilot. ;) Its conclusion is 'Overall, the code appears to be safe and follows good practices for unit testing in Python.'

WPK4FEM reviewed (dismissed):
Dear Anne,
I don't have comments here that others haven't already raised ( or not even those ). It remains scary to me that a single test changes its environment and has to restore the environment. However that is in a place we do not directly control and you have documented what and why.
Nothing further, I'm happy that now this test will be executed regularly.
Wijtze Pieter

avdg81 committed:
Processed the review suggestions

- Renamed a function object.
- Clarified a comment.
- Removed an unnecessary slicing operation
- Used a better way to remove the previous test directory when the test is set up

rfaasse reviewed (approved):
Thanks for processing the changes, it's good to go!
